doc/man3/SSL_CTX_set_new_pending_ssl_cb.pod

```
@@ -0,0 +1,48 @@
+=pod
+
+=head1 NAME
+
+SSL_CTX_set_new_pending_ssl_cb - callback function to report creation of QUIC connection SSL objects 
+
+=head1 SYNOPSIS
+
+ typedef void (*SSL_set_new_pending_ssl_cb_fn)(SSL_CTX *c, SSL *new_ssl,
+                                               void *arg);
+ void SSL_CTX_set_new_pending_ssl_cb(SSL_CTX *c,
+                                     SSL_set_new_pending_ssl_cb_fn *f,
+                                     void *arg);
+
+=head1 DESCRIPTION
+
+SSL_CTX_set_new_pending_ssl_cb() sets the new_pending_ssl callback function and
+associated application data argument B<arg>.  When using the QUIC transport, TLS
+handshake processing may occur independently from the thread which accepts the
+connection that the handshake is establishing.  As such, SSL objects
+representing the connection may be allocated and initialized prior to a call to
+SSL_accept_connection().  This callback is mean to be informative to an
+application that a new SSL object is being created, and will soon be available
+in the accept queue.
+
+=head1 RETURN VALUES
+
+Neither SSL_CTX_set_new_pending_ssl_cb() nor the registered callback function
+return any value.
```

Should it be possible for the callback to abort? Let's say some fatal error occurs (e.g. malloc failure) in the callback, would it be beneficial to be able to say...abort what you're doing we can't accept this, e.g. via the callback returning 0.

---

I had taken this callback to be informative, rather than controlling.  For example, if this callback is made, and some application level failure prevents it from doing whatever it wants to do, that doesn't mean that the QUIC stack can't further process the new connection (that is to say, the channel and user_ssl object is already allocated and configured).  If there is a desire to limit the number of connections an application wants to allow in the backlog, I can certainly understand that, but I think the nominal method for doing that would be to restrict the length on the ports incoming_channel_list, and expose that control via an API (i.e. something like SSL_set_max_backlog(), functioning in a similar fashion to what /proc/sys/net/ipv4/tcp_max_syn_backlog does).  That way, we can reject subsequent connection requests in the port handler without having to iteratively do an allocation/callback/rejection/free cycle (working under the assumption that even a transient failure in the callback will be repeated under high incoming connection situations).

---

But is there any reason *not* to allow this level of control? Essentially we leave the decision to the app. Is this failure sufficiently serious that we should stop processing the request, or should we continue anyway? It seems like this could be quite useful and easy for us to do.

---

This is a tough question for me to answer.  I can't argue the point that its easy to do, its is quite simple.  I think what I'm struggling with are the repercussions of doing so.  What we're talking about here is the following sequence of events:

1) QUIC stack creates a new channel and tls object
2) QUIC stack makes a callback to inform the application of (1)
3) Application determines for whatever reason, that it can't handle a new connection, and returns an error to the stack
4) QUIC stack frees the channel and associated tls object

Whats bothering me is that this creates what appears to me to be a fairly heavy processing loop.  i.e. This callback gets called every time a new channel is created.  Ostensibly a common failure pattern would be an overloaded system due to a large number of QUIC connection requests.  So in that situation we would potentially be doing steps 1-4 iteratively for a large number of iterations (i.e. an alloc/init/callback/free) for every new incomming connection, which seems inefficient to me, hence my comment about creating a incomming queue limiting api.

Thats not to say we can't have both (this now and a queue limiter later), but I'm not sure that we want that.  If you're ok with both, then I'm ok with adding this, I'm just hesitant.

---

> I think what I'm struggling with are the repercussions of doing so

I guess I'm worried about this from the opposite angle, i.e. what do applications do if they can't do this? If an application relies on all SSL objects being decorated with some ex_data, and suddenly they have a situation where they can't do so (doesn't have to be malloc failure, could be some other problem completely unrelated to an overloaded system). This could have major repercussions for that application. Maybe they assume that the ex_data will always be there and dereference it - but now they have to review the entire app and allow the possibility that an SSL will exist undecorated....and probably that will just lead to a failure in the application even further down the processing chain.

So it my mind it is much better to fail early at the source of the problem.

---

reading my comment out loud, I just went ahead and made the change.  I personally don't think I'd use this call for the reasons above, but I'm having a hard time seeing why we couldn't have this and a queue limiting call at the same time